### ISSUE TITLE
Linux Python issue

### ISSUE THREAD

**Issue by ObiSandwich** [team-hunting/ComicDownloader#3]
OS: Linux LMDE4
Python 3.7.3

Unfortunately can't get your new script to run :( My system is saying Python is up to date, but getting the errors below. So putting this out there now while I look into it to document and to see if any one else has a fix in the mean time.

`./comicScraper.py https://readcomiconline.li/Comic/The-Superior-Foes-of-Spider-Man/Issue-7?id=2109
/usr/lib/python3/dist-packages/requests/__init__.py:91: RequestsDependencyWarning: urllib3 (1.26.7) or chardet (3.0.4) doesn't match a supported version!
  RequestsDependencyWarning)
ComicScraper v0.1.14 
You have provided the link for a single issue.
When providing the URL for the info page, this script will download all issues in the series.
No title specified. Reading title from Url...
Using title: The-Superior-Foes-of-Spider-Man-Issue-7
Downloading max quality images
Starting to scrape The-Superior-Foes-of-Spider-Man-Issue-7 from https://readcomiconline.li/Comic/The-Superior-Foes-of-Spider-Man/Issue-7?id=2109
Number of Issues to download 1
Issues to download: 
['&readType=1']
Downloading Issue &readType=1
Traceback (most recent call last):
  File "./comicScraper.py", line 607, in <module>
    main(downloadFull, singleIssue, comicTitle, lowres, disableWait, startURL, useSelenium, seleniumDisplay, issueStart)
  File "./comicScraper.py", line 482, in main
    imageLinks, issueImageDict = downloadAllWithRequests(fullComicDownload, startURL, issueLinks, title, singleIssueDownload, disableWait)
  File "./comicScraper.py", line 438, in downloadAllWithRequests
    path = saveImagesFromImageLinks(issueImageDict[issueName], len(issueImageDict[issueName]), issueName)
KeyError: '&readType=1'
`

Beautifulsoup error...

`sudo pip install -r requirements.txt
Collecting beautifulsoup4==4.10.0 (from -r requirements.txt (line 1))
  Could not find a version that satisfies the requirement beautifulsoup4==4.10.0`

**Comment by TravisHunting** [comment]
Sounds like our requirements.txt isn't up to date... will take a look.
You can try running "pip install beautifulsoup4"

**Comment by AustinHunting** [comment]
in relation to the pip error:

It appears to me as though you might be using pip from python 2.7 as that version of `beautifulsoup4` is not available in the PyPi-2 repository anymore. Can you post the output of a few commands for use so we can verify what environment this is being run in? 

Please run and post the outputs of:
```
python3 --version

pip -V

pip freeze
```

Lets see if we can get the `pip` issue sorted out and before tackling the `KeyError: '&readType=1'` issue.

**Comment by ObiSandwich** [comment]
Thanks, but I did try this one before...

`$pip install beautifulsoup4`

`WARNING: pip is being invoked by an old script wrapper. This will fail in a future version of pip.
Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.`

`DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. pip 21.0 will drop support for Python 2.7 in January 2021. More details about Python 2 support in pip can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support pip 21.0 will remove support for this functionality.`

`Defaulting to user installation because normal site-packages is not writeable
Requirement already satisfied: beautifulsoup4 in ./.local/lib/python2.7/site-packages (4.9.3)
Requirement already satisfied: soupsieve<2.0,>1.2; python_version < "3.0" in ./.local/lib/python2.7/site-packages (from beautifulsoup4) (1.9.6)
Requirement already satisfied: backports.functools-lru-cache; python_version < "3" in ./.local/lib/python2.7/site-packages (from soupsieve<2.0,>1.2; python_version < "3.0"->beautifulsoup4) (1.6.4)`

**Comment by ObiSandwich** [comment]
`$ python3 --version`
`Python 3.7.3`

`$ pip -V`
`WARNING: pip is being invoked by an old script wrapper. This will fail in a future version of pip.
Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.
pip 20.3.4 from /home/aaa/.local/lib/python2.7/site-packages/pip (python 2.7)`

`$ pip freeze`
`WARNING: pip is being invoked by an old script wrapper. This will fail in a future version of pip.
Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.`

`DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. pip 21.0 will drop support for Python 2.7 in January 2021. More details about Python 2 support in pip can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support pip 21.0 will remove support for this functionality.`
alabaster==0.7.8
asn1crypto==0.24.0
audioread==2.1.5
Babel==2.6.0
backports.functools-lru-cache==1.6.4
beautifulsoup4==4.9.3
certifi==2018.8.24
chardet==3.0.4
colorama==0.4.4
configobj==5.0.6
configparser==3.5.0b2
crayons==0.4.0
cryptography==2.6.1
docutils==0.14
entrypoints==0.3
enum34==1.1.6
html5lib==1.0.1
idna==2.6
imagesize==1.0.0
ipaddress==1.0.17
Jinja2==2.10
keyring==17.1.1
keyrings.alt==3.1.1
lxml==4.3.2
MarkupSafe==1.1.0
mutagen==1.40.0
mysqlclient==1.3.10
numpy==1.16.2
olefile==0.46
packaging==19.0
Pillow==5.4.1
puddletag==1.2.0
pyacoustid==1.1.5
pycairo==1.16.2
pycrypto==2.6.1
Pygments==2.3.1
PyGObject==3.30.4
pymad==0.10
pyOpenSSL==19.0.0
pyparsing==2.2.0
python-apt==1.8.4.3
python-Levenshtein==0.12.0
python-xlib==0.23
pytz==2019.1
pyxdg==0.25
requests==2.21.0
roman==2.0.0
SecretStorage==2.3.1
selenium==3.141.0
setproctitle==1.1.10
six==1.12.0
soupsieve==1.9.6
Sphinx==1.8.4
sphinx-rtd-theme==0.4.3
typing==3.6.6
urllib3==1.24.1
webdriver-manager==2.4.0
webencodings==0.5.1

**Comment by ObiSandwich** [comment]
Sorry to bring my Linux sh1t to the party, but Mint has been pretty good to me over the last few years, not really required Python before until I found your magic script! Be interesting to know if any else one has had any similar issues on a non-Mint Linux?

**Comment by AustinHunting** [comment]
No worries on the linux problems, we're building this script on Ubuntu 20.04 and windows 10. 
Our script requires python3 to run, we ensure that it's being used with the shebang at the start of the file. Unfortunately there's not a single command to ensure a `pip install` runs with pip or pip3. It seems as though Mint comes with both python 3 and python 2.7 installed, and the default pip package is for python 2.7.

Looks like the `pip` issue you're seeing is that python 2.7's pip is installing the packages and python3 is trying to use them.
You'll need to run `pip3 install -r requirements.txt` if that fails with something to the affect of `Command 'pip3' not found...` you'll need to install it with `sudo apt update && sudo apt install python3-pip` and then rerun the `pip3 install -r requirements.txt`. 

To summarize, run:
```
which pip3 || (sudo apt update && sudo apt install python3-pip -y) && pip3 install -r requirements.txt
```
And let's see if that fixes the issue.

**Comment by ObiSandwich** [comment]
Thanks for this, I think this worked: `pip3 install -r requirements.txt`

Looked different than before and there was definitely something installing this time - here's the end of the output;

`The script normalizer is installed in '/home/aaa/.local/bin' which is not on PATH.
Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed beautifulsoup4-4.10.0 bs4-0.0.1 certifi-2021.10.8 charset-normalizer-2.0.10 idna-3.3 requests-2.27.0 soupsieve-2.3.1`

But, run this one again: `./comicScraper.py https://readcomiconline.li/Comic/The-Superior-Foes-of-Spider-Man/Issue-7?id=2109`

Again, output different to original one above, but still no download :( Are there any clues in any of this?

`ComicScraper v0.1.14`

`You have provided the link for a single issue.
When providing the URL for the info page, this script will download all issues in the series.
No title specified. Reading title from Url...
Using title: The-Superior-Foes-of-Spider-Man-Issue-7
Downloading max quality images
Starting to scrape The-Superior-Foes-of-Spider-Man-Issue-7 from https://readcomiconline.li/Comic/The-Superior-Foes-of-Spider-Man/Issue-7?id=2109
Number of Issues to download 1`

`Issues to download: 
['&readType=1']
Downloading Issue &readType=1
Traceback (most recent call last):
  File "./comicScraper.py", line 607, in <module>
    main(downloadFull, singleIssue, comicTitle, lowres, disableWait, startURL, useSelenium, seleniumDisplay, issueStart)
  File "./comicScraper.py", line 482, in main
    imageLinks, issueImageDict = downloadAllWithRequests(fullComicDownload, startURL, issueLinks, title, singleIssueDownload, disableWait)
  File "./comicScraper.py", line 438, in downloadAllWithRequests
    path = saveImagesFromImageLinks(issueImageDict[issueName], len(issueImageDict[issueName]), issueName)
KeyError: '&readType=1'
`

**Comment by TravisHunting** [comment]
Hm yeah that definitely doesn't seem right

**Comment by TravisHunting** [comment]
Alright, I've merged in some updates. 
The Single Issue downloads should work again. I've also made a change to the way it detects the actual image URL when scraping, so theoretically this should be able to scrape those 'corrupt' jpgs we talked about before.

Side note, we recommend using with the -s flag for selenium, it seems to be more accurate when finding the images. It also is confirmed to work with any images with abnormal file names, like the star wars ones you wanted before.

**Comment by ObiSandwich** [comment]
That worked! Thanks so much - That Spidey Single worked,  "Corrupt" JPGs also looks good. But I did get the following error on this, but all 2,070 images downloaded ok any way, so not sure what the error was;

`./comicScraper.py https://readcomiconline.li/Comic/Star-Wars-2015 -c`

`Number of images to download: 
2070
Downloading images...
Downloaded 2070 files to /home/aaa/Downloads/- COMICS/ComicDownloader-main/Star-Wars-2015//Complete/
Total size of files: 2390.699mb
Average file size: 1.155mb
All files are larger than 10kb - good to go!
There was an error downloading ['Star-Wars-2015-Complete']
Downloaded:
Comics have been moved to Star-Wars-2015/CBZ_Files`

I tried the following, but nothing much happened other than the output below;

`./comicScraper.py https://readcomiconline.li/Comic/Star-Wars-2015 -s`

Starting Selenium...

====== WebDriver manager ======
Could not get version for google-chrome with the any command: google-chrome --version || google-chrome-stable --version
Current google-chrome version is UNKNOWN
Get LATEST chromedriver version for UNKNOWN google-chrome
Driver [/home/aaa/.wdm/drivers/chromedriver/linux64/97.0.4692.71/chromedriver] found in cache

Traceback (most recent call last):
  File "./comicScraper.py", line 626, in <module>
    main(downloadFull, singleIssue, comicTitle, lowres, disableWait, startURL, useSelenium, seleniumDisplay, issueStart)
  File "./comicScraper.py", line 499, in main
    imageLinks, issueImageDict = downloadAllWithSelenium(fullComicDownload, startURL, issueLinks, title, singleIssueDownload, disableWait, seleniumDisplay)
  File "./comicScraper.py", line 424, in downloadAllWithSelenium
    driver = webdriver.Chrome(service=s, options=options)
  File "/home/aaa/.local/lib/python3.7/site-packages/selenium/webdriver/chrome/webdriver.py", line 73, in __init__
    service_log_path, service, keep_alive)
  File "/home/aaa/.local/lib/python3.7/site-packages/selenium/webdriver/chromium/webdriver.py", line 99, in __init__
    options=options)
  File "/home/aaa/.local/lib/python3.7/site-packages/selenium/webdriver/remote/webdriver.py", line 268, in __init__
    self.start_session(capabilities, browser_profile)
  File "/home/aaa/.local/lib/python3.7/site-packages/selenium/webdriver/remote/webdriver.py", line 359, in start_session
    response = self.execute(Command.NEW_SESSION, parameters)
  File "/home/aaa/.local/lib/python3.7/site-packages/selenium/webdriver/remote/webdriver.py", line 424, in execute
    self.error_handler.check_response(response)
  File "/home/aaa/.local/lib/python3.7/site-packages/selenium/webdriver/remote/errorhandler.py", line 247, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: unknown error: cannot find Chrome binary
Stacktrace:
#0 0x55bda9154a23 <unknown>
#1 0x55bda8c1fe18 <unknown>
#2 0x55bda8c4114b <unknown>
#3 0x55bda8c3e91a <unknown>
#4 0x55bda8c7974a <unknown>
#5 0x55bda8c73883 <unknown>
#6 0x55bda8c493fa <unknown>
#7 0x55bda8c4a4c5 <unknown>
#8 0x55bda918416d <unknown>
#9 0x55bda919a5bb <unknown>
#10 0x55bda9185e75 <unknown>
#11 0x55bda919ae85 <unknown>
#12 0x55bda917986f <unknown>
#13 0x55bda91b5ae8 <unknown>
#14 0x55bda91b5c68 <unknown>
#15 0x55bda91d0aad <unknown>
#16 0x7f07d695afa3 <unknown>

**Comment by TravisHunting** [comment]
Thanks for all your input.
That error is completely erroneous. It fires every time LOL, haven't bothered to fix it. It's because the cbz file isn't getting moved into the right folder.
"There was an error downloading ['xxxxx']"

**Comment by TravisHunting** [comment]
If you get this message:
All files are larger than 10kb - good to go! 

It should mean that there were zero corrupted images downloaded. (in theory)

**Comment by ObiSandwich** [comment]
Hey no worries and thanks for explaining that... but, as I couldn't get the `-s` argument to work, what is it supposed to do? Could you add a command example to the ReadMe please?
